src/clientiface.cpp

```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
-		wanted_range);
-
-	s16 d_max = full_d_max;
-
-	// Don't loop very much at a time
-	s16 max_d_increment_at_time = 2;
-	if (d_max > d_start + max_d_increment_at_time)
-		d_max = d_start + max_d_increment_at_time;
-
-	// cos(angle between velocity and camera) * |velocity|
-	// Limit to 0.0f in case player moves backwards.
-	f32 dot = rangelim(camera_dir.dotProduct(playerspeed), 0.0f, 300.0f);
-
-	// Reduce the field of view when a player moves and looks forward.
-	// limit max fov effect to 50%, 60% at 20n/s fly speed
-	camera_fov = camera_fov / (1 + dot / 300.0f);
-
-	s32 nearest_emerged_d = -1;
-	s32 nearest_emergefull_d = -1;
-	s32 nearest_sent_d = -1;
-	//bool queue_is_full = false;
-
-	const v3s16 cam_pos_nodes = floatToInt(camera_pos, BS);
-
-	s16 d;
-	for (d = d_start; d <= d_max; d++) {
-		/*
-			Get the border/face dot coordinates of a "d-radiused"
-			box
-		*/
-		std::vector<v3s16> list = FacePositionCache::getFacePositions(d);
-
-		std::vector<v3s16>::iterator li;
-		for (li = list.begin(); li != list.end(); ++li) {
-			v3s16 p = *li + center;
-
-			/*
-				Send throttling
-				- Don't allow too many simultaneous transfers
-				- EXCEPT when the blocks are very close
-
-				Also, don't send blocks that are already flying.
-			*/
-
-			// Start with the usual maximum
-			u16 max_simul_dynamic = max_simul_sends_usually;
-
-			// If block is very close, allow full maximum
-			if (d <= BLOCK_SEND_DISABLE_LIMITS_MAX_D)
-				max_simul_dynamic = m_max_simul_sends;
-
-			// Don't select too many blocks for sending
-			if (num_blocks_selected >= max_simul_dynamic) {
-				//queue_is_full = true;
-				goto queue_full_break;
-			}
-
-			// Don't send blocks that are currently being transferred
-			if (m_blocks_sending.find(p) != m_blocks_sending.end())
+	for (size_t i=0; i<m_map_send_queue.size(); i++) {
+		const WantedMapSend &wms = m_map_send_queue[i];
+		if (wms.type == WMST_MAPBLOCK) {
+			/*verbosestream << "Server: Client "<<peer_id<<" wants MapBlock ("
+					<<wms.p.X<<","<<wms.p.Y<<","<<wms.p.Z<<")"
+					<< std::endl;*/
+
+			// Do not go over-limit
+			if (blockpos_over_max_limit(wms.p))
 				continue;
 
-			/*
-				Do not go over max mapgen limit
-			*/
-			if (blockpos_over_max_limit(p))
+			// Don't send blocks that are currently being transferred
+			if (m_blocks_sending.find(wms) != m_blocks_sending.end())
 				continue;
 
-			// If this is true, inexistent block will be made from scratch
-			bool generate = d <= d_max_gen;
-
-			/*
-				Don't generate or send if not in sight
-				FIXME This only works if the client uses a small enough
-				FOV setting. The default of 72 degrees is fine.
-				Also retrieve a smaller view cone in the direction of the player's
-				movement.
-				(0.1 is about 4 degrees)
-			*/
-			f32 dist;
-			if (!(isBlockInSight(p, camera_pos, camera_dir, camera_fov,
-						d_blocks_in_sight, &dist) ||
-					(playerspeed.getLength() > 1.0f * BS &&
-					isBlockInSight(p, camera_pos, playerspeeddir, 0.1f,
```

Comment:
Will we still get the extra blocks in player-move-direction? I put that in a while ago, so that a player does not fall into an ungenerated map. (Especially if the player is looking into a different direction)

Reply:
I moved that part to an earlier stage in the process. My version changes the entirety of p, camera_pos, camera_dir and camera_fov to be a block ahead and pointing at the movement direction if the player or the object the player is attached to is moving at considerable speed. Only if the player is stationary or roughly walking do they represent the actual position and camera direction of the player. The exact behavior that this should have is a big question mark but it should be fine for the purpose it's serving at the moment.